src/Core/View.js

```diff
@@ -107,8 +107,21 @@ function _preprocessLayer(view, layer, provider) {
             layer.tileTextureCount = provider.tileTextureCount.bind(provider);
         }
 
-        if (provider.preprocessDataLayer) {
-            provider.preprocessDataLayer(layer);
+        if (!layer.preprocessedPromise) {
+            if (provider.preprocessDataLayer) {
+                const preprocessingResult = provider.preprocessDataLayer(layer, view, view.mainLoop.scheduler);
+                if (preprocessingResult && preprocessingResult.then) {
+                    layer.preprocessedPromise = preprocessingResult;
+                } else {
+                    layer.preprocessedPromise = Promise.resolve();
+                }
+            } else {
+                layer.preprocessedPromise = Promise.resolve();
+            }
+            layer.preprocessedPromise.then(() => {
+                layer.ready = true;
```

Comment:
The alternative is to wrap every code that needs this boolean in whenReady.then callback (if they accept to run in another microtask).